File: DD_c_plusplus/Examples/SOCKnapsackInstance/SOCKnapsackInstance.cpp
#include "SOCKnapsackInstance.h"


SOCKnapsackInstance::SOCKnapsackInstance(const string& file_path) {
    read_file(file_path);
}

SOCKnapsackInstance::SOCKnapsackInstance(const SOCKnapsackInstance& other)
    : omega(other.omega)
    , right_side_of_restrictions(other.right_side_of_restrictions)
    , objective_weights(other.objective_weights)
    , linear_weights(other.linear_weights)
    , sqrt_weights(other.sqrt_weights)
    , variables(other.variables)
    , initial_state(other.initial_state ? new vector<double>(*other.initial_state) : nullptr)
{}

SOCKnapsackInstance& SOCKnapsackInstance::operator=(const SOCKnapsackInstance& other) {
    if (this == &other) return *this;
    delete initial_state;
    omega = other.omega;
    right_side_of_restrictions = other.right_side_of_restrictions;
    objective_weights = other.objective_weights;
    linear_weights = other.linear_weights;
    sqrt_weights = other.sqrt_weights;
    variables = other.variables;
    initial_state = other.initial_state ? new vector<double>(*other.initial_state) : nullptr;
    return *this;
}

void SOCKnapsackInstance::read_file(const string& file_path) {
    ifstream input(file_path);
    if (!input.is_open()) {
        cerr << "Error: could not open file " << file_path << endl;
        exit(1);
    }

    int variable_length, num_constraints;
    input >> variable_length >> num_constraints;

    input >> omega;

    // Read objective coefficients, one per variable
    for (int i = 0; i < variable_length; i++) {
        double value;
        input >> value;
        objective_weights.push_back(value);
    }

    // Read right-hand side, one per constraint
    for (int i = 0; i < num_constraints; i++) {
        int value;
        input >> value;
        right_side_of_restrictions.push_back(value);
    }

    // Read linear weight matrix: one row per constraint, one column per variable
    for (int i = 0; i < num_constraints; i++) {
        vector<int> row;
        for (int j = 0; j < variable_length; j++) {
            int value;
            input >> value;
            row.push_back(value);
        }
        linear_weights.push_back(row);
    }

    // Read square-root weight matrix: one row per constraint, one column per variable
    for (int i = 0; i < num_constraints; i++) {
        vector<int> row;
        for (int j = 0; j < variable_length; j++) {
            int value;
            input >> value;
            row.push_back(value);
        }
        sqrt_weights.push_back(row);
    }

    for (int i = 1; i <= variable_length; ++i) {
        string variable_name = "x_" + to_string(i);
        variables.emplace_back(make_pair(variable_name, vector<int>{0, 1}));
    }

    initial_state = new vector<double>{0.0, 0.0};
}

File: DD_python/Examples/SOCKnapsack/SOCKnapsackInstance.py
class SOCKnapsackInstance:
    def __init__(self, file_path: str) -> None:
        """
        SourceCode to store the structure of the knapsack problem.
        """

        self.initial_state: list[int] = None
        self.variables: list[tuple[str, list[int]]] = None
        self.right_side_of_restrictions: list[int] = None
        self.linear_weights: list[list[int]] = []
        self.sqrt_weights: list[list[int]] = []
        self.omega: int = None
        self.objective_weights: list[int] = None


        self.read_file(file_path)

    def read_file(self, file_path: str) -> None:
        try:
            with open(file_path, 'r') as input_file:
                [variable_length, num_constraints] = list(map(int, input_file.readline().strip().split()))
                self.omega = int(input_file.readline().strip())

                self.objective_weights = list(map(int, input_file.readline().strip().split()))
                self.right_side_of_restrictions = list(map(int, input_file.readline().strip().split()))

                for i in range(num_constraints):
                    self.linear_weights.append(list(map(int, input_file.readline().strip().split())))

                for i in range(num_constraints):
                    self.sqrt_weights.append(list(map(int, input_file.readline().strip().split())))

                self.variables = [("x_" + str(i + 1), [0, 1]) for i in range(variable_length)]
                self.initial_state = [0, 0]

        except IOError:
            print(f"Error: could not open file {file_path}")
            exit(1)

    def __str__(self):
        num_vars = len(self.variables) if self.variables else 0
        num_constraints = len(self.linear_weights)

        return (
            f"SOCKnapsackInstance\n"
            f"-------------------\n"
            f"Variables (n): {num_vars}\n"
            f"Constraints (m): {num_constraints}\n"
            f"Omega: {self.omega}\n\n"
            
            f"Initial State:\n{self.initial_state}\n\n"
            
            f"Objective Weights:\n{self.objective_weights}\n\n"
            
            f"Right-hand Side (b):\n{self.right_side_of_restrictions}\n\n"
            
            f"Linear Weights (A):\n"
            + "\n".join([str(row) for row in self.linear_weights]) + "\n\n"
            
            f"Quadratic Weights (Q):\n"
            + "\n".join([str(row) for row in self.sqrt_weights])
        )